app/src/main/java/com/jerboa/ui/components/blocks/BlocksActivity.kt

```diff
@@ -0,0 +1,260 @@
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.platform.LocalContext
+import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.text.TextStyle
+import androidx.compose.ui.text.font.FontWeight
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.unit.sp
+import com.jerboa.R
+import com.jerboa.api.API
+import com.jerboa.api.ApiState
+import com.jerboa.api.toApiState
+import com.jerboa.model.SiteViewModel
+import com.jerboa.showBlockCommunityToast
+import com.jerboa.showBlockInstanceToast
+import com.jerboa.showBlockPersonToast
+import com.jerboa.ui.components.common.ApiEmptyText
+import com.jerboa.ui.components.common.ApiErrorText
+import com.jerboa.ui.components.common.CircularIcon
+import com.jerboa.ui.components.common.JerboaSnackbarHost
+import com.jerboa.ui.components.common.LoadingBar
+import com.jerboa.ui.components.common.SimpleTopAppBar
+import com.jerboa.ui.components.common.simpleVerticalScrollbar
+import com.jerboa.ui.theme.Title
+import it.vercruysse.lemmyapi.v0x19.datatypes.BlockCommunity
+import it.vercruysse.lemmyapi.v0x19.datatypes.BlockCommunityResponse
+import it.vercruysse.lemmyapi.v0x19.datatypes.BlockInstance
+import it.vercruysse.lemmyapi.v0x19.datatypes.BlockInstanceResponse
+import it.vercruysse.lemmyapi.v0x19.datatypes.BlockPerson
+import it.vercruysse.lemmyapi.v0x19.datatypes.BlockPersonResponse
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+
+private const val UNBLOCK_BUTTON_SIZE = 18
+
+@Composable
+private fun BlockedElementListItem(
+    apiState: ApiState<*>,
+    icon: String?,
+    name: String,
+    onUnblock: () -> Unit,
+    onSuccessfulUnblock: () -> Unit
+) {
+    Row(verticalAlignment = Alignment.CenterVertically) {
+        icon?.let {
+            CircularIcon(
+                icon = it,
+                contentDescription = "",
+                size = 26.dp,
+            )
+            Spacer(modifier = Modifier.padding(horizontal = 4.dp))
+        }
+        Text(name, modifier = Modifier.weight(1f))
+        TextButton(onClick = onUnblock, colors= ButtonDefaults.buttonColors(Color.Transparent)) {
+            when (apiState) {
+                ApiState.Loading -> CircularProgressIndicator(
+                    modifier = Modifier.size(UNBLOCK_BUTTON_SIZE.dp),
+                    color = Color.Gray,
+                )
+                is ApiState.Success -> onSuccessfulUnblock()
+                else -> Text(
+                    text = "X",
+                    style = TextStyle(
+                        color = Color.Red,
+                        fontWeight = FontWeight.Bold,
+                        fontSize = UNBLOCK_BUTTON_SIZE.sp,
+                    ),
+                )
+            }
+        }
+    }
+}
+
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+fun BlocksActivity(
+    siteViewModel: SiteViewModel,
```

Comment:
I would make a viewModel for this screen. Keep the UI and data logic separate.

Reply:
Absolutely, but I'd also add the block actions to the SiteViewModel rather than creating a new one, so that it can update the blocks. We may also need to change / fix some other block actions, such as `CommunityViewModel.blockCommunity`, so that it can stay in sync with `SiteViewModel.my_user.community_blocks`... etc.

Reply:
@dessalines I don't think I understand the idea of adding the block actions to the SiteViewModel. I had to create an apiState for each element on the list to avoid the same state being displayed to every element (this would cause the loading state to be displayed on each element, for instance).

Reply:
Add what you can to the blockviewmodel. I don't understand  what you would need a apistate per block

Reply:
@MV-GH you suggest a list of ApiStates inside the BlockViewModel? That's interesting

Reply:
Well or a BlockListViewModel and have a viewmodel per entry.

Reply:
See these guidelines on why we do this.


https://developer.android.com/develop/ui/compose/architecture#udf
https://android.googlesource.com/platform/frameworks/support/+/androidx-main/compose/docs/compose-api-guidelines.md#prefer-stateless-and-controlled-composable-functions